File: dictionary.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
	"unicode"
)

type Dictionary struct {
	size int
	cap  int
	data []string
}

func NewDictionary() *Dictionary {
	return &Dictionary{
		size: 0,
		cap:  10,
		data: make([]string, 10),
	}
}

func (d *Dictionary) AddWord(str string) {
	if d.size == d.cap {
		newCap := d.cap * 2
		newData := make([]string, newCap)
		copy(newData, d.data)
		d.data = newData
		d.cap = newCap
	}
	d.data[d.size] = str
	d.size++
}

func (d *Dictionary) Errors(str string) int {
	result := 0
	words := strings.Fields(str)

	for _, word := range words {
		wordLower := strings.ToLower(word)

		// Если слово полностью в нижнем регистре
		if word == wordLower {
			result++
			continue
		}

		// Проверка на несколько заглавных букв
		manyUpper := false
		upperCount := 0
		for _, ch := range word {
			if unicode.IsUpper(ch) {
				upperCount++
				if upperCount > 1 {
					manyUpper = true
					break
				}
			}
		}
		if manyUpper {
			result++
			continue
		}

		// Проверяем есть ли слово в словаре (без учета регистра)
		haveWord := false
		for i := 0; i < d.size; i++ {
			dictWordLower := strings.ToLower(d.data[i])
			if wordLower == dictWordLower {
				haveWord = true
				break
			}
		}

		// Проверяем точное совпадение
		noError := false
		for i := 0; i < d.size; i++ {
			if word == d.data[i] {
				noError = true
				break
			}
		}

		if haveWord && !noError {
			result++
		}
	}
	return result
}

func (d *Dictionary) String() string {
	return strings.Join(d.data[:d.size], " ") + "\n"
}

func dictionary() {
	fmt.Println("Введите словарь (слова через пробел), затем слово 'stop', затем строку для проверки:")

	dict := NewDictionary()
	scanner := bufio.NewScanner(os.Stdin)

	if !scanner.Scan() {
		return
	}
	input := scanner.Text()

	words := strings.Fields(input)
	foundStop := false
	var textToCheck string

	for _, word := range words {
		if word == "stop" {
			foundStop = true
			break
		}
		dict.AddWord(word)
	}

	if foundStop {
		// Извлекаем оставшийся текст после 'stop'
		stopIndex := strings.Index(input, "stop")
		if stopIndex != -1 {
			textToCheck = strings.TrimSpace(input[stopIndex+4:])
		}
	}

	// Если текст для проверки пустой, читаем новую строку
	if textToCheck == "" {
		fmt.Println("Введите строку для проверки:")
		if !scanner.Scan() {
			return
		}
		textToCheck = scanner.Text()
	}

	// Проверяем ошибки
	errorCount := dict.Errors(textToCheck)

	fmt.Printf("Словарь: %s", dict.String())
	fmt.Printf("Проверяемая строка: %s\n", textToCheck)
	fmt.Printf("Количество ошибок: %d\n", errorCount)
}

File: dictionary.cpp
#include <iostream> //cAnnot cannOt fOund pAge stop thE pAge cAnnot be found - ввод проверки
#include <algorithm>
#include <sstream>

using namespace std;

struct Dictionary{
    size_t size;
    size_t cap;
    string* data;
    Dictionary() : size(0), cap(10)
                , data(new string[10]){}
    void addWord(string str){
        data[size] = str;
        size++;
        if (size == cap){
            string* newdata = new string[2*cap];
            cap *= 2;
            for(int i = 0; i < size; i++) newdata[i] = data[i];
            delete data;
            data = newdata;
        }
    }

    int errors(string str){
        int result = 0;
        stringstream ss(str);
        while (ss >> str){
            string str_down = str;
            transform(str.begin(), str.end(), str_down.begin(), ::tolower); //нет ударений
            if (str_down == str){
                result++;
                continue;
            }

            bool manyUdr = false;
            for(size_t i = 0, udr = 0 ; i < str.size(); i++){
                    if (str[i] != str_down[i]) udr++;
                    if (udr > 1) {
                        manyUdr = true;
                        break;
                    }
                }
            if (manyUdr){
                result++;
                continue;
            }
            bool haveWord = false; //есть ли слово в принципе
            for(int i = 0; i < size; i++){
                string data_down = data[i];
                transform(data[i].begin(), data[i].end(), data_down.begin(), ::tolower);
                if (str_down == data_down){
                    haveWord = true;
                };
            }

            bool noError = false;
            for(int i = 0; i < size; i++){
                if (str == data[i]){
                    noError = true;
                };
            }

            if (haveWord && !(noError)) result++;
            
        }
        return result;
    }
    

    ~Dictionary(){
        delete[] data;
    }
};

ostream& operator<<(ostream& ss, Dictionary& dic){
    for(int i = 0; i < dic.size; i++) ss << dic.data[i] << " ";
    ss << endl;
    return ss;
}

void dictionary(){
    cout << "Введите словарь (слова через пробел), затем слово 'stop', затем строку для проверки:" << endl;
    
    Dictionary dict;
    string input;
    
    cin.ignore();
    getline(cin, input);
    
    stringstream ss(input);
    string usCin;
    
    while(ss >> usCin){
        if (usCin == "stop"){
            break;
        }
        dict.addWord(usCin);
    }
    
    string text_to_check;
    if (getline(ss, text_to_check)) {
        // Убираем лишние пробелы в начале
        text_to_check.erase(0, text_to_check.find_first_not_of(" "));
    }
    
    // Если в строке не осталось текста для проверки, читаем новую строку
    if (text_to_check.empty()) {
        cout << "Введите строку для проверки:" << endl;
        getline(cin, text_to_check);
    }
    
    // Проверяем ошибки
    int error_count = dict.errors(text_to_check);
    
    
    cout << "Словарь: " << dict;
    cout << "Проверяемая строка: " << text_to_check << endl;
    cout << "Количество ошибок: " << error_count << endl;
}